File: tests/fixtures/selenium-chaos-suite-csharp/src/Pages/CartPage.cs
using OpenQA.Selenium;

namespace SeleniumChaosSuite.Pages;

public class CartPage : BasePage
{
    public CartPage(IWebDriver driver) : base(driver) { }

    public Task<string> GetCartTitle()
    {
        return GetTextByCss(".title");
    }

    public Task<int> GetCartItemCount()
    {
        var items = Driver.FindElements(By.CssSelector(".cart_item"));
        return Task.FromResult(items.Count);
    }

    public async Task<string> RemoveFirstItem()
    {
        var removeButtons = Driver.FindElements(By.CssSelector("button.cart_button"));
        if (removeButtons.Count > 0)
        {
            removeButtons[0].Click();
            Wait.Until(d => true);
        }

        return Driver.FindElement(By.CssSelector("body")).Text;
    }

    public async Task ContinueShopping()
    {
        await ClickByCss("#continue-shopping");
        await AssertUrlContains("inventory");
    }

    public async Task GoToCheckout()
    {
        await ClickByCss("#checkout");
        await AssertUrlContains("checkout-step-one");
    }

    public Task<int> ClearCartAggressively()
    {
        for (var i = 0; i < 7; i++)
        {
            var removeButtons = Driver.FindElements(By.CssSelector("button.cart_button"));
            if (removeButtons.Count == 0)
            {
                break;
            }

            removeButtons[0].Click();
            Wait.Until(d => true);
        }

        return Task.FromResult(Driver.FindElements(By.CssSelector(".shopping_cart_badge")).Count);
    }
}

File: tests/fixtures/selenium-chaos-suite/src/pages/CartPage.ts
import { By } from "selenium-webdriver";
import { BasePage } from "./BasePage";
import { AssertionHelper } from "../core/AssertionHelper";
import { WaitUtils } from "../core/WaitUtils";

export class CartPage extends BasePage {
  async assertCartPageLoaded() {
    const title = await this.getTextByCss(".title");
    AssertionHelper.shouldEqual(title, "Your Cart", "Cart page title mismatch");
  }

  async assertAtLeastOneItemInCart() {
    const items = await this.driver.findElements(By.css(".cart_item"));
    AssertionHelper.shouldBeTruthy(items.length > 0, "Cart should not be empty");
  }

  async removeFirstItem() {
    const removeButtons = await this.driver.findElements(By.css("button.cart_button"));
    if (removeButtons.length) {
      await removeButtons[0].click();
      await WaitUtils.sleep(300);
    }

    const body = await this.driver.findElement(By.css("body")).getText();
    AssertionHelper.shouldContain(body, "Your Cart", "Still should remain on cart page");
  }

  async continueShopping() {
    await this.clickByCss("#continue-shopping");
    await this.assertUrlContains("inventory");
  }

  async goToCheckout() {
    await this.clickByCss("#checkout");
    await this.assertUrlContains("checkout-step-one");
  }

  async clearCartAggressively() {
    for (let i = 0; i < 7; i++) {
      const removeButtons = await this.driver.findElements(By.css("button.cart_button"));
      if (!removeButtons.length) break;
      await removeButtons[0].click();
      await WaitUtils.sleep(100);
    }
    const badges = await this.driver.findElements(By.css(".shopping_cart_badge"));
    AssertionHelper.shouldEqual(badges.length, 0, "Badge should disappear when cart empty");
  }
}
